lenskit-core/src/main/java/org/lenskit/data/entities/AbstractBeanEntity.java

```diff
@@ -0,0 +1,192 @@
+     * Construct a bean entity.
+     */
+    protected AbstractBeanEntity(EntityType typ, long id) {
+        super(typ, id);
+        Pair<AttributeSet,ImmutableList<Method>> res = lookupAttrs(getClass());
+        attributes = res.getLeft();
+        methods = res.getRight();
+    }
+
+    @Override
+    public Set<TypedName<?>> getTypedAttributeNames() {
+        return attributes;
+    }
+
+    @Override
+    public Set<String> getAttributeNames() {
+        return attributes.nameSet();
+    }
+
+    @Override
+    public boolean hasAttribute(String name) {
+        return attributes.nameSet().contains(name);
+    }
+
+    @Override
+    public boolean hasAttribute(TypedName<?> name) {
+        return attributes.contains(name);
+    }
+
+    @Nullable
+    @Override
+    public Object maybeGet(String attr) {
+        int idx = attributes.lookup(attr);
+        if (idx >= 0) {
+            Method mh = methods.get(idx);
+            try {
+                return mh.invoke(this);
+            } catch (IllegalAccessException | InvocationTargetException e) {
+                throw new RuntimeException("Error invoking " + mh, e);
+            }
+        } else {
+            return null;
+        }
+    }
+
+    @Nullable
+    @Override
+    public <T> T maybeGet(TypedName<T> name) {
+        int idx = attributes.lookup(name, true);
+        if (idx >= 0) {
+            Method mh = methods.get(idx);
+            try {
+                return (T) mh.invoke(this);
+            } catch (IllegalAccessException | InvocationTargetException e) {
+                throw new RuntimeException("Error invoking " + mh, e);
+            }
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public long getLong(TypedName<Long> name) {
+        int idx = attributes.lookup(name);
+        if (idx >= 0) {
+            Method mh = methods.get(idx);
+            try {
+                return (long) mh.invoke(this);
+            } catch (IllegalAccessException | InvocationTargetException e) {
+                throw new RuntimeException("Error invoking " + mh, e);
+            }
+        } else {
+            throw new NoSuchAttributeException(name.toString());
+        }
+    }
+
+    @Override
+    public double getDouble(TypedName<Double> name) {
+        int idx = attributes.lookup(name);
+        if (idx >= 0) {
+            Method mh = methods.get(idx);
+            try {
+                return (double) mh.invoke(this);
+            } catch (IllegalAccessException | InvocationTargetException e) {
+                throw new RuntimeException("Error invoking " + mh, e);
+            }
+        } else {
+            throw new NoSuchAttributeException(name.toString());
+        }
+    }
+
+    @Override
+    public int getInteger(TypedName<Integer> name) {
+        int idx = attributes.lookup(name);
+        if (idx >= 0) {
+            Method mh = methods.get(idx);
+            try {
+                return (int) mh.invoke(this);
```

Comment:
That is... a good question. A very good question.

For these abstract bean entities, it'll box (I think). For the basic entity, things are boxed. I would like to make the abstract bean entities not box, but that *looks* like it will require codegen.

For  the packed entity storage I am currently building, I think I'll be able to do it without boxing.

Not sure if it's worth it.